File: src/de/ax/powermode/MyCaretListener.scala
package de.ax.powermode

import com.intellij.openapi.editor.Caret
import com.intellij.openapi.editor.event.{CaretEvent, CaretListener}

import scala.util.Try

/**
  * Created by nyxos on 04.01.17.
  */
class MyCaretListener extends CaretListener with Power {
  var modified = true


  override def caretPositionChanged(caretEvent: CaretEvent): Unit = {
    if (!modified && powerMode.caretAction) {
      initializeAnimationByCaretEvent(caretEvent.getCaret)
    }
    modified = false
  }

  override def caretRemoved(caretEvent: CaretEvent): Unit = {
    modified = true
  }

  override def caretAdded(caretEvent: CaretEvent): Unit = {
    modified = true
  }

  private def initializeAnimationByCaretEvent(caret: Caret) {
    val isActualEditor = Try {
      caret.getEditor.getColorsScheme.getClass.getName.contains("EditorImpl")
    }.getOrElse(false)
    if (isActualEditor) {
      Util.getCaretPosition(caret).toOption.foreach(p => powerMode.maybeElementOfPowerContainerManager.foreach(_.initializeAnimation(caret.getEditor, p)))
    }
  }
}

File: src/com/nmeylan/powermode/MyCaretListener.java
package com.nmeylan.powermode;

import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.event.CaretEvent;
import com.intellij.openapi.editor.event.CaretListener;
import org.jetbrains.annotations.NotNull;

import java.awt.Point;

class MyCaretListener implements CaretListener, Power {
  private boolean modified = true;

  @Override
  public void caretPositionChanged(@NotNull CaretEvent event) {
    if (!modified && powerMode().isCaretActionEnabled()) {
      initializeAnimationByCaretEvent(event.getCaret());
    }
    modified = false;
  }

  @Override
  public void caretAdded(@NotNull CaretEvent event) {
    modified = true;
  }

  @Override
  public void caretRemoved(@NotNull CaretEvent event) {
    modified = true;
  }

  private void initializeAnimationByCaretEvent(Caret caret) {
    if (Util.isActualEditor(caret.getEditor())) {
      Point position = Util.getCaretPosition(caret);
      powerMode().getMaybeElementOfPowerContainerManager().ifPresent(e -> e.initializeAnimation(caret.getEditor(), position));
    }
  }
}
